include/dsn/tool-api/network.h

```diff
@@ -271,6 +271,10 @@ class rpc_session : public ref_counter
     ::dsn::utils::ex_lock_nr _lock; // [
     volatile session_state _connect_state;
 
+    // when a session isn't connected, all messages are queued in _pending_connected.
+    // after connected, all of them are moved to "_messages"
+    std::vector<message_ex *> _pending_connected;
```

Comment:
Why introduce it? Suppose the negotiation procedure last a very long time, `_pending_connected` may consume a large amount of memory.

Reply:
So what's your opinion? throw all of these requests which are generated during negotiation?

Reply:
I think drop them is OK, message in the queue may also timeouted when negotiation finished.
However, I want to hear somebody other's voice.

Reply:
@neverchanje @hycdong

Reply:
I agree with @acelyc111. It's an early-optimization.

Actually the client could also have implemented a pending queue. The requests may be queued on the client-side until the session authenticated. Thus the server-queue during authentication could always be empty and useless.

Reply:
I agree with you. But if we remove the `pending_connected`, so the cpp sdk will not support the pending queue

Reply:
There's always one pending queue on the client-side once it supports asynchronous operations.